drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +45,494 @@ extern "C" {
+    _vector_table_0: [u32; 6],          // 0x08, 0c, 10, 14, 18, 1c
+    nxp_image_length: u32,              // 0x20
+    nxp_image_type: u32,                // 0x24
+    nxp_offset_to_specific_header: u32, // 0x28
+    _vector_table_1: [u32; 2],          // 0x2c, 0x30
+    nxp_image_executation_address: u32, // 0x32
+                                        // Additional trailing vectors are not
+                                        // interesting here.
+                                        // _vector_table_2[u32; 2]
+}
+
+impl ImageVectorsLpc55 {
+    const IMAGE_TYPE_PLAIN_SIGNED_XIP_IMAGE: u32 = 4;
+
+    pub fn is_image_type_signed_xip(&self) -> bool {
+        self.nxp_image_type == Self::IMAGE_TYPE_PLAIN_SIGNED_XIP_IMAGE
+    }
 
     // This part aliases flash in two positions that differ in bit 28. To allow
     // for either position to be used in new images, we clear bit 28 in all of
     // the numbers used for comparison below, by ANDing them with this mask:
-    const ADDRMASK: u32 = !(1 << 28);
+    pub fn normalized_initial_pc(&self) -> u32 {
+        const ADDRMASK: u32 = !(1 << 28);
+        self.initial_pc & ADDRMASK
+    }
+
+    // Length of image from offset zero to end of the signature block (without padding)
+    // a.k.a. ImageVectorsLpc55.nxp_image_length
+    pub fn image_length(&self) -> Option<u32> {
+        if self.is_image_type_signed_xip() {
+            Some(self.nxp_image_length)
+        } else {
+            None
+        }
+    }
+
+    /// Image length padded to nearest page size.
+    pub fn padded_image_len(&self) -> Option<u32> {
+        round_up_to_flash_page(self.image_length()?)
+    }
+
+    /// Determine the bounds of an image assuming the given flash bank
+    /// addresses.
+    pub fn padded_image_range(&self, exec: &Range<u32>) -> Option<Range<u32>> {
+        let image_start = exec.start;
+        let image_end = exec.start.saturating_add(self.padded_image_len()?);
+        // ImageHeader is optional. Assuming code cannot be earlier than
+        // IMAGE_HEADER_OFFSET.
+        let initial_pc_start = image_start.saturating_add(IMAGE_HEADER_OFFSET);
+        let initial_pc_end =
+            image_start.saturating_add(self.nxp_offset_to_specific_header);
+        let pc = exec.start.saturating_add(self.normalized_initial_pc());
+
+        if !exec.contains(&image_end)
+            || !exec.contains(&initial_pc_start)
+            || !exec.contains(&initial_pc_end)
+            || !(initial_pc_start..initial_pc_end).contains(&pc)
+        {
+            // One of these did not fit in the destination flash bank or
+            // exeecutable area of the image in that bank.
+            return None;
+        }
+        Some(image_start..image_end)
+    }
+}
+
+impl TryFrom<&[u8]> for ImageVectorsLpc55 {
+    type Error = ();
+
+    fn try_from(buffer: &[u8]) -> Result<Self, Self::Error> {
+        match ImageVectorsLpc55::read_from_prefix(buffer) {
```

Comment:
Are you comfortable with the fact that this will only work on 32-bit aligned u8 arrays? If so, it needs a comment to that effect.

Reply:
read_from_prefix copies into an aligned struct.

```
$ cat src/main.rs 
use anyhow::*;
use zerocopy::*;

#[derive(Debug, FromBytes, FromZeroes, Default)]
#[repr(C)]
struct Msg {
    n: u32,
    b: u16,
}

fn main() {
    let bytes: &[u8] = &[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    println!("data: {:?}", bytes);

    for start in 0..(bytes.len() - core::mem::size_of::<Msg>()) {
        let msg = Msg::read_from_prefix(&bytes[start..])
            .ok_or_else(|| anyhow!("failed to parse Msg"))
            .unwrap();
        println!("msg @{}: {:#x?}", start, msg);
    }
}
$ cargo run
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/struct-from-bytes`
data: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
msg @0: Msg {
    n: 0x3020100,
    b: 0x504,
}
msg @1: Msg {
    n: 0x4030201,
    b: 0x605,
}
msg @2: Msg {
    n: 0x5040302,
    b: 0x706,
}
```

Reply:
Ah, my mistake, I confused it with the one that does a direct ref. This is making a copy.